fix: Parse Relation not working

YBTopaz8 opened the pull request:
Closes: #397

Did some general enhancements on the code structure it's more consistent (Working on `object` types are cool but a pain to debug since you never know the real type until you hit an exception.)

YBTopaz8 commented:
@mtrezza to help you out,
I did ZERO new addition to the code. I simply forked multiple times and fixed the specific bug/scope.
If you need an order through which you should merge, consider this as the First One to approve.

mtrezza commented:
You mentioned in the PR title only Parse Relations. But according to the description this PR closes [3 bugs](https://github.com/parse-community/Parse-SDK-dotNET/issues/397#issue-2751384561); which one is true?

YBTopaz8 commented:
@mtrezza  they are all fixed !
Usernames are now returned as expected 
And saving work on users, objects, relations - in fact all MutableObjects should all be saved properly.

This [sample repo of mine](https://github.com/YBTopaz8/Chat-App-.NET-MAUI-LIVE-QUERY) showcases most if not all the basic/core Parse operations + having Parse Live Queries methods too for anyone in need of a Demo.

Those issues we caused by encoding, like with User, operations (and querying)

mtrezza commented:
Nice, but how is it possible that the CI passes here even before we merged https://github.com/parse-community/Parse-SDK-dotNET/pull/403?

YBTopaz8 commented:
Because Like I said, the main issues were very like likely linked to either encoding or the way we received Https responses.

So fixing "Encoding" likely provided a stable foundation for all Https Enc/Decoding.

That's my thought process at least.

YBTopaz8 commented:
So they likely pass because Relations didn't have methods implemented AND didn't have test implemented to test coverage.

YBTopaz8 commented:
In anticipation to any other questions let me fully explain the changes I made...;
1. Changes All tests especially the latest ones as they lacked proper ini and and clean up. (Some test descriptions we added here and there
2. Parse Relations were broken due to no update to the encoding in order to support relations. As they now inherited from IParseFieldOperation which is JSON convertible, I had to ensure operations are converted and de-converted as needed. 
3. ParseCommand, Data which gets the command data, encodes in Stream to POST/PUT had an issue where the operation was only done once and out of sync compared to its context. So I had to put a null coalescing operator as an check that should usually ensure we have the object with its data on demand.
4. There are some other changes I made like breaking down 1 line into 2 because writing one line of code having over 4 methods chained together is never helpful especially after pouring hours on understanding the code base.


These are the highlights. any other change was just to ensure field operations work as they should.

mtrezza commented:
Could you please take a look at the conflicts so we can review this?

mtrezza commented on Parse.Tests/ConfigTests.cs:
```diff
@@ -115,14 +117,16 @@ await Assert.ThrowsExceptionAsync<TaskCanceledException>(async () =>
-            Assert.ThrowsException<KeyNotFoundException>(() => configuration.Get<string>("doesNotExist"));
-        }
+
+        //[TestMethod]
```
Can this commented out block be removed?

mtrezza replied:
removed

mtrezza commented on Parse/Infrastructure/Data/ParseDataDecoder.cs:
```diff
@@ -20,44 +20,44 @@ public class ParseDataDecoder : IParseDataDecoder
+            IDictionary<string, object> { } dictionary when dictionary.ContainsKey("__op") => ParseFieldOperations.Decode(dictionary),
+
+            IDictionary<string, object> { } dictionary when dictionary.TryGetValue("__type", out var type) && Types.Contains(type) => type switch
             {
```
Could you please revert all code style changes in code that is unrelated to this issue. It seems that there hasn't been any code change other than the indentation, haven't looked closely though.

Can we solve this with a linter? Is there a linter already in place for this repo?

YBTopaz8 replied:
I, do not know what *linters* would be but I believe [this](https://github.com/parse-community/Parse-SDK-dotNET/pull/393#issuecomment-2535314741) would be related.

For the code itself, would you mean I would; remove these and submit in a different PR?

mtrezza replied:
Not sure I follow. I was referring to changes that are only related to spacing for example. If they are not in code that is related to the issue the PR addresses, they should be revered. An example is the file I tagged above, there seem to be no code changes, only indentation or line-break changes. These changes should be reverted in this PR. They don't need to be added in a separate PR, a linter should take care of that.

YBTopaz8 replied:
Oh. I'm afraid my knowledge is quite sparse when it comes to "lints" then..

mtrezza replied:
See https://github.com/parse-community/Parse-SDK-dotNET/pull/404#pullrequestreview-2523532572

mtrezza reviewed (approved):
As I've pointed out in https://github.com/parse-community/Parse-SDK-dotNET/pull/404#discussion_r1897055356 this PR contains a lot of unrelated formatting changes like indentation, spaces, newline. This makes the PR more difficult to review. Some of these changes are likely done manually, other may be done automatically by your IDE. Since we don't have a linter yet, we'll make an exception and ignore these changes in this PR, so we can go ahead faster. However, to make PRs easier to review, could you please make sure that a PR doesn't contain such changes? Just take a look at the changes before you the commit. If you see any unnecessary line changes, please revert them before committing. You can also see these changes on GitHub in the "Files changed" tab. I've opened https://github.com/parse-community/Parse-SDK-dotNET/pull/406 to add a linter, which will prevent these changes to get merged in the future, but it's still in the works.

YBTopaz8 commented:
Sure thing. I'll be more diligent on that aspect.

Adding the linter will help indeed, as sometimes the file changes don't even reflect on my end.

But all is good now. 

Next PR should be easier to Navigate through

mtrezza commented:
@YBTopaz8 Before I merge, could you please suggest a title for this PR that describes all the things that this PR fixes? The https://github.com/parse-community/Parse-SDK-dotNET/issues/397 seems quite broad as it mentions 3 problems in https://github.com/parse-community/Parse-SDK-dotNET/issues/397#issue-2751384561 and in https://github.com/parse-community/Parse-SDK-dotNET/issues/397#issuecomment-2558176172 you mention that it "clears your issues and any other potential one" so this PR fixes more than the 3 issues? If yes, what are these other issues? This is quite confusing.

YBTopaz8 commented:
Sure.
The title is quite fitting you may add (+ bug fixes)
As I 
1. Fixed `.SaveAsync()` which didn't work due to encoder/decoder not accounting for all possible data types we support
2. `ParseClient.Instance.GetCurrentUser();` now returns all the expected data including User name (was missing before)
3. All Relations-Related issues are fixed (Get, Save, Remove)
4. `ParseClient.Instance.Logout()` can now be called on its own, without async nor cancellation token. If in need to pass a token, we can still call `await ParseClient.Instance.LogoutAsync()`. The aim was to avoid doing everything asynchronously
5. Changed from `IsAuthenticated` Property to a method `IsAuthenticatedAsync()` both returning a bool, but the latter was needed to correct `.SaveAsync()`
6. Misc change: renamed `LoginAsync()` to `LoginWithAsync()` and `SignUpAsync()` to `SignUpWithAsync()`. Just a name change as per comments left in the file - and it makes it more intuitive so I did that too.
7. Changes done to UWC : redid my approach as explained already - ensuring Data transfer is posted and gotten correctly. 

That's pretty much it. everything else is unchanged (unless I haven't caught it/if it's an issue!)
 
Just ONE change I want to do later and it will be to REPLACE this below. It doesn't negatively affect At all. but it will always print out "Unsupported length" all the time as an exception. I have the fix already but will hold on to that, perhaps in the next PR - depends on you I suppose.

```
    try
        {
            totalLength = responseStream.Length;
        }
        catch
        {
            Console.WriteLine("Unsupported length...");
        };
```

mtrezza commented:
Are all these points 1-7 related to the bug with Parse Relations? Or are these actually 7 individual bugs?

YBTopaz8 commented:
- All of these are ultimately related to Encoding/Decoding due to v4.0.0's implementation being incomplete at the time.
Point 7 is an adjacent change aimed at making code more **readable**. All of the changes in Point 7 can be reverted if need.

- None of the bugs here are new since Parse supported the features before. 
- None of these are aimed optimizing performance..

- For breaking Changes.. Well on top of more tests, I Created a repo, now online, where I tested [the core Parse Functionalities (login, save file, CRUD parse objects and Relations, call cloud code function, and much more!)](https://github.com/YBTopaz8/Chat-App-.NET-MAUI-LIVE-QUERY/tree/master) And they all work and are reproducible to/by all. 

- Any bug occurring would be an interesting one to me as well and I'll be sure to look into

mtrezza commented:
I don't follow. I was asking whether these points are related to the issue with Parse Relation and you say they are all related to encoding/decoding. So how is encoding/decoding related to the bug with Parse Relation?

I also don't understand: "None of the bugs here are new since Parse supported the features before." So this PR does fix other bugs besides the Parse Relation bug?

mtrezza commented:
@YBTopaz8 How's it going with this PR? What do you thin is needed to merge it, or parts of it?

parseplatformorg commented:
🎉 This change has been released in version [4.0.2](https://github.com/parse-community/Parse-SDK-dotNET/releases/tag/4.0.2)
